File: backend/repositories/event_repository.py
"""Persistence operations for events."""

from typing import Any

from config.dynamodb import dynamodb_resource
from config.environment import EVENTS_TABLE


class EventRepository:
    def __init__(self, table: Any | None = None) -> None:
        self.table = table or dynamodb_resource().Table(EVENTS_TABLE)

    def list_events(self) -> list[dict[str, Any]]:
        items: list[dict[str, Any]] = []
        response = self.table.scan()
        items.extend(response.get("Items", []))
        while response.get("LastEvaluatedKey"):
            response = self.table.scan(ExclusiveStartKey=response["LastEvaluatedKey"])
            items.extend(response.get("Items", []))
        return items

    def get_event(self, event_id: str) -> dict[str, Any] | None:
        return self.table.get_item(Key={"event_id": event_id}).get("Item")

File: backend/repositories/eventRepository.js
/**
 * Persistence operations for events.
 */

const { ScanCommand, GetCommand } = require("@aws-sdk/lib-dynamodb");
const { dynamoDBDocumentClient } = require("../config/dynamodb");
const { EVENTS_TABLE } = require("../config/environment");

class EventRepository {
  /**
   * @param {Object} [options]
   * @param {Object} [options.client] - DynamoDB Document Client override (for testing).
   */
  constructor({ client } = {}) {
    this.client = client || dynamoDBDocumentClient();
    this.tableName = EVENTS_TABLE;
  }

  /**
   * List all events from the table (handles pagination).
   *
   * @returns {Promise<Object[]>}
   */
  async listEvents() {
    const items = [];
    let lastKey;

    do {
      const response = await this.client.send(
        new ScanCommand({
          TableName: this.tableName,
          ...(lastKey && { ExclusiveStartKey: lastKey }),
        })
      );
      items.push(...(response.Items || []));
      lastKey = response.LastEvaluatedKey;
    } while (lastKey);

    return items;
  }

  /**
   * Get a single event by ID.
   *
   * @param {string} eventId
   * @returns {Promise<Object|null>}
   */
  async getEvent(eventId) {
    const response = await this.client.send(
      new GetCommand({
        TableName: this.tableName,
        Key: { event_id: eventId },
      })
    );
    return response.Item || null;
  }
}

module.exports = { EventRepository };
